File: python/tests/test_a2a_server.py
"""Tests for A2A server setup."""

import pytest
from timestep.a2a.server import create_agent_card, create_server


def test_create_skills_from_tools():
    """Test creating AgentSkill objects from tools."""
    from timestep.a2a.postgres_agent_store import Agent
    
    # Create a test agent with both tools
    test_agent = Agent(
        id="test-agent",
        name="Test Agent",
        description="Test agent",
        tools=["get_weather", "web_search"],
        model="gpt-4.1",
    )
    
    # Test by checking the agent card skills
    card = create_agent_card(test_agent)
    skills = card.skills
    
    assert len(skills) == 2
    
    # Find skills by ID (order may vary)
    weather_skill = next((s for s in skills if s.id == "get_weather"), None)
    search_skill = next((s for s in skills if s.id == "web_search"), None)
    
    assert weather_skill is not None
    assert weather_skill.name == "Get Weather"
    assert len(weather_skill.examples) > 0
    
    assert search_skill is not None
    assert search_skill.name == "Web Search"
    assert len(search_skill.examples) > 0


def test_create_agent_card():
    """Test creating AgentCard."""
    from timestep.a2a.postgres_agent_store import Agent
    
    url = "http://localhost:8080/"
    test_agent = Agent(
        id="test-agent",
        name="Test Agent",
        description="Test agent",
        tools=["get_weather", "web_search"],
        model="gpt-4.1",
    )
    
    card = create_agent_card(test_agent, url)
    
    assert card.name == "Test Agent"
    assert card.url == url
    assert card.version == "2026.0.5"
    assert card.default_input_modes == ["text"]
    assert "text" in card.default_output_modes
    assert "task-status" in card.default_output_modes
    assert card.capabilities.streaming is True
    assert len(card.skills) == 2
    # Examples may not be a direct attribute, check if it exists
    if hasattr(card, "examples"):
        assert len(card.examples) > 0


def test_create_server():
    """Test creating A2A server."""
    app = create_server(host="127.0.0.1", port=8080)
    
    # Verify server was created (should be a Starlette app)
    assert app is not None
    # Check that it has routes (Starlette apps have routes)
    assert hasattr(app, "routes") or hasattr(app, "router")


def test_create_server_with_custom_tools():
    """Test creating server with custom tools."""
    from timestep.core import GetWeather
    
    app = create_server(host="127.0.0.1", port=8080, tools=[GetWeather])
    assert app is not None


def test_create_server_with_custom_model():
    """Test creating server with custom model."""
    app = create_server(host="127.0.0.1", port=8080, model="gpt-4")
    assert app is not None


def test_create_agent_card_with_default_url():
    """Test creating agent card with default URL."""
    from timestep.a2a.postgres_agent_store import Agent
    
    test_agent = Agent(
        id="test-agent",
        name="Test Agent",
        description="Test agent",
        tools=["get_weather"],
        model="gpt-4.1",
    )
    
    card = create_agent_card(test_agent)
    assert card.url == "http://localhost:8080/"

File: typescript/tests/test_a2a_server.ts
/** Tests for A2A server setup. */

import { describe, it, expect, vi } from 'vitest';

// Mock the pg module (native module that vitest can't load)
vi.mock('pg', () => ({
  default: {
    Pool: vi.fn().mockImplementation(() => ({
      on: vi.fn(),
      connect: vi.fn(),
      query: vi.fn(),
      end: vi.fn(),
    })),
  },
  Pool: vi.fn().mockImplementation(() => ({
    on: vi.fn(),
    connect: vi.fn(),
    query: vi.fn(),
    end: vi.fn(),
  })),
}));

import { createAgentCard, createServer } from '../timestep/a2a/server.js';
import { GetWeatherParameters } from '../timestep/core/tools.js';
import type { Agent } from '../timestep/a2a/postgres_agent_store.js';

describe('A2A Server Tests', () => {
  it('should create agent card with correct properties', () => {
    /** Test creating AgentCard. */
    const url = 'http://localhost:8080/';
    const testAgent: Agent = {
      id: 'test-agent',
      name: 'Test Agent',
      description: 'Test agent',
      tools: ['get_weather', 'web_search'],
      model: 'gpt-4.1',
      created_at: new Date(),
      updated_at: new Date(),
    };
    const card = createAgentCard(testAgent, url);

    expect(card.name).toBe('Test Agent');
    expect(card.url).toBe(url);
    expect(card.version).toBe('2026.0.5');
    expect(card.defaultInputModes).toEqual(['text']);
    expect(card.defaultOutputModes).toContain('text');
    expect(card.defaultOutputModes).toContain('task-status');
    expect(card.capabilities?.streaming).toBe(true);
    expect(card.skills).toBeDefined();
    expect(card.skills?.length).toBe(2);
    // Examples may not be a direct attribute, check if it exists
    const cardWithExamples = card as unknown as { examples?: string[] };
    if (cardWithExamples.examples) {
      expect(cardWithExamples.examples.length).toBeGreaterThan(0);
    }
  });

  it('should create agent card with skills', () => {
    /** Test that agent card includes skills. */
    const testAgent: Agent = {
      id: 'test-agent',
      name: 'Test Agent',
      description: 'Test agent',
      tools: ['get_weather', 'web_search'],
      model: 'gpt-4.1',
      created_at: new Date(),
      updated_at: new Date(),
    };
    const card = createAgentCard(testAgent);
    
    expect(card.skills).toBeDefined();
    expect(card.skills?.length).toBe(2);
    
    const weatherSkill = card.skills?.find(s => s.id === 'get_weather');
    expect(weatherSkill).toBeDefined();
    expect(weatherSkill?.name).toBe('Get Weather');
    
    const searchSkill = card.skills?.find(s => s.id === 'web_search');
    expect(searchSkill).toBeDefined();
    expect(searchSkill?.name).toBe('Web Search');
  });

  it('should create server application', () => {
    /** Test creating A2A server. */
    const app = createServer('127.0.0.1', 8080);
    
    expect(app).toBeDefined();
    // Express app should have methods like use, listen, etc.
    expect(typeof app.listen).toBe('function');
  });

  it('should create server with custom tools', () => {
    /** Test creating server with custom tools. */
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const tools = [
      { name: 'get_weather', parameters: GetWeatherParameters },
    ];
    const app = createServer('127.0.0.1', 8080, tools);
    expect(app).toBeDefined();
  });

  it('should create server with custom model', () => {
    /** Test creating server with custom model. */
    const app = createServer('127.0.0.1', 8080, undefined, 'gpt-4');
    expect(app).toBeDefined();
  });

  it('should create agent card with default URL', () => {
    /** Test creating agent card with default URL. */
    const testAgent: Agent = {
      id: 'test-agent',
      name: 'Test Agent',
      description: 'Test agent',
      tools: ['get_weather'],
      model: 'gpt-4.1',
      created_at: new Date(),
      updated_at: new Date(),
    };
    const card = createAgentCard(testAgent);
    expect(card.url).toBe('http://localhost:8080/');
  });
});

